src/persist/src/operators/stream.rs

```diff
@@ -739,114 +742,31 @@ where
-
-                while let Some((cap, error_cap, pending_future)) = pending_futures.front_mut() {
-                    match Pin::new(pending_future).poll(&mut context) {
-                        std::task::Poll::Ready(result) => {
-                            match result {
-                                Ok(seq_no) => {
-                                    log::trace!(
-                                        "In {}, finished writing for time: {}, seq_no: {:?}",
-                                        &operator_name,
-                                        cap.time(),
-                                        seq_no,
-                                    );
-                                }
-                                Err(e) => {
-                                    let mut session = error_output.session(&error_cap);
-                                    let error = format!(
-                                        "In {}, error writing data for time {}: {}",
-                                        &operator_name,
-                                        error_cap.time(),
-                                        e
-                                    );
-                                    log::error!("{}", error);
-
-                                    // TODO: make error retractable? Probably not...
-                                    session.give((error, *error_cap.time(), 1));
-                                }
-                            }
-
-                            let _ = pending_futures.pop_front().expect("known to exist");
-                        }
-                        std::task::Poll::Pending => {
-                            // We assume that write requests are worked off in order and stop
-                            // trying for the first write that is not done.
-                            break;
-                        }
-                    }
-                }
-            }
-        });
-
-        (data_output_stream, error_output_stream)
+        let (pass_through, to_retract) =
+            self.branch(move |_, (_, t, _)| if t < &upper_ts { false } else { true });
+
+        let (retract_oks, errs) = to_retract
+            .map(|(data, time, diff)| (data, time, -diff))
+            .persist(name, write);
+
+        // Introduce a data-dependency between pass_through, and retract_oks even
+        // though in reality they operate on disjoint subsets of data and no true
+        // data dependency exists. This way, we ensure that the frontier doesn't
+        // advance for downstreams users until after all retractions have been
+        // persisted.
+        //
+        // TODO: we could have done this with fewer operators by concatenating
+        // pass_through, to_retract, and retract_oks, as to_retract and retract_oks
+        // cancel out. This approach seemed more isolated and safer.
+        let retract_oks = retract_oks.map(|_| None);
+
+        let oks = pass_through
```

Comment:
note to self: if the extra operators end up being too slow, we could write a `and_then` operator that takes two inputs A and B and passes through data from A, and maintains capabilities for the min of (input frontier of A, input frontier of B). Then we wouldn't need to do this map / filter / map dance to go to an Option and back

Reply:
I figured out a better solution -- flat_map!